trunk/Annotator/resources/man/man1/OnsetExtractor.1
.TH ONSETEXTRACTOR "1" "December 2008" "OnsetExtractor" "User Commands"
.SH NAME
OnsetExtractor \- Detects starting notes in a wave file
.SH SYNOPSIS
.B OnsetExtractor
[\fIoptions\fR] \fIaudiofile1 audiofile2 \fR...
.SH DESCRIPTION
OnsetExtractor extracts the onsets of a wave file containing music,
and exports it as CLAM Annotator XML format.
An onset is the starting point of a note or a percusion instrument.
.SH OPTIONS
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-s\fR SCHEMAFILE, \fB\-\-print\-schema\fR=\fISCHEMAFILE\fR
Outputs the description schema for the extractor. Use
\&'\-' for standard output.
.TP
\fB\-w\fR, \fB\-\-writeback\fR
Writes back the descriptors to the source. (Not
applied in this extrator)
.TP
\fB\-f\fR SUFFIX, \fB\-\-suffix\fR=\fISUFFIX\fR
Appends SUFFIX to the generated descriptors file
(default: '.pool')
.TP
\fB\-c\fR CONFIGURATIONFILE, \fB\-\-configuration\fR=\fICONFIGURATIONFILE\fR
Specifies a configuration file with the options for
the extractor.
.SH "SEE ALSO"
Annotator(1), ClamExtractorExample(1)



CLAM/CHANGES


2008-??-?? CLAM 1.3.1 SVN $Revision$
''
 * Typed controls: Controls with other type than floats.
 * Networks can pause without stopping the backend.
 * Processing::Get{In,Out}{Control,Port}() overloaded for indexes and names.
 * Processing::GetN{In,Out}{Controls,Ports}() to get the number of connectors.
 * Processing::Get{In,Out}{Controls,Ports}() are now protected methods.

2008-08-11 CLAM 1.3.0
'Shooting flying plugins release'
 * SVN Revision: 11960.
 * Integrates first wave of GSoC 2008 contributions from the students
   Natanael Olaiz, Wang Jun, Pawel Bartkiewicz, Francisco Tufro,
   and Yushen Han.
 * Native windows compilation (using mingw) working without patches (GSoC:jun)
 * A simpler and more effective API to export LADSPA plugins from networks
   and isolated processings.
 * Template based processing/plugin code generator scripts and frontend.
 * Many changes to the Network allowing to do cut and paste, embeded
   positions...
 * Spatialization plugin:
   * ComplexSpectrumInverter: Inverts the phase of a ComplexSpectrum
   * Click-less impulse response switching,
   * Choreography sequencer to calculate simulated impulses on the fly
     instead of using a dense grid of precomputed IRs.
     (requires a non-public binary but you can plug-in a different one)
   * Spherical2CartesianControl: Converts radix,azimut,elevation controls
     to x,y,z
 * Sample by sample processings now are a plugin.
 * New processings:
   * AudioFileMemoryLoader: an audio file reader without locks and with seek
     but loading the full audio into memory (a ringbuffer based on the way).
   * Resampler: Resamples audio streams using libsamplerate (not stable)
 * TickExtractor example working again (thanks 
 * Fix: NetEditor Windows installer should not install .def and .lib for
   plugins
 * Voice2MIDI crosscompiled windows binaries (no installer yet)
 * Old AudioFileInput/Output definitely removed, remaining old code adapted.
 * FormantTracker and TickExtractor examples working with current version

2008-02-05 CLAM 1.2.0
'GSoCket Plugged-in release'
 * SVN Revision: 11171
 * Major changes summary:
   * Integrates GSoC students work. Thanks to Hernan Ordiales
      Bennet Kolasinsky, Greg Kellum, Andreas Calvo, Roman Goj,
      Abe Kazemzadeh and Google Inc.
   * It also integrates work done last month within the
     Acoustics team on the Barcelona Media center.
   * CLAM plugins: You can load your own processings and
     data types in an existing CLAM application such the
     Network Editor. (GSoC:acalvo)
   * Windows binaries crosscompile from Linux
 * NetworkEditor enhancements:
   * Enhanced online tutorial
   * Token data types expose a color to show in NetworkEditor
   * Port tooltips show also the data type
   * Suitable monitors are available at port context menu
   * Suitable controlers are available at control context menu
     or double clicking
   * The processing is automatically populated with available
     processings. (GSoC:acalvo)
   * Dialog to add processings has completion
   * New spectrogram view (GSoC:bkolasinsky)
 * Realtime port to some spectral processings and prototypes:
    * Harmonizer, pitch-discretization and morph (GSoC:hordia)
 * Chord analysis modularity, configurability and postprocessing
   improved (GSoC:rgoj)
 * Anotator enhancements (GSoC:bkolasinsky):
   * Multiple segmentation and low-level features panes
   * BarGraph is available as Instant View
   * Rulers are positioned at round numbers
   * Rulers have subticks for smaller not numbered divisions
   * Schema browsers shows Scopes (ie. Frame) as subtrees of
     their defining attribute (ie. Song::FrameDivision)
 * First plugins available:
   * guitardistortion: Cool guitar effects with prototyper
     interface. (GSoC:hordia)
   * continuousExcitationSynth: An SMS based synthesizer.
     (GSoC:gkellum)
   * speech: Experimental processings for speech synthesis
     and analysis (GSoC:akazemzadeh)
   * spacialization: 3D audio processing including
     * New spectral processings with simplifyied/explicit
       polar-complex representation, including windowing,
       fft, ifft, overlap-and-add and representation
       converters.
     * Real-time low latency convolution
     * Grid and polar impulse response database fetcher
     * Exibition processings to convert p+v format to
       5:1, N-Surround and other exibition formats.
     * Several room databases are available, check the wiki.
 * Build system:
   * Windows binaries obtained by crosscompiling from linux
   * Windows external libs are deployed now as a posix tree
   * Relying on pkg-config to link dependencies
   * VisualC bits droped
   * Installing plugins
   * OSX uses macported Qt instead of Trolltech binaries

2007-06-07 CLAM 1.1.0
'More eye-candy, please'
 * SVN Revision: 10216
 * New concise task oriented documentation (on the wiki)
 * Loading of faust (ladspa) modules
 * FLTK fully dropped
 * Qt3 based vmqt dropped (some files moved to SMSTools)
 * '_support' ending for scons configure options removed:
   with_jack_support, with_ladspa_support, with_osc_support
   became with_jack, with_ladspa, with_osc
 * Examples clean up
 * Realtime port to some processings:
   * SMSGenderChange, SMSPitchShift, SMSShapeShift
 
2007-03-20 CLAM 1.0.0
'The Berlin release'
 * SVN Revision: 9881
 * Funtional Ladspa host

2007-03-18 CLAM 0.99.0
'That will be CLAM 1.0 release'
 * SVN Revision: 9864
 * New binary in NetworkEditor/ called OfflinePlayer to batch-process
   clam network files given an input and output audio file.
 * Fixes in the modconf to work in suse (thanks Toni)
 * Updated Fedora Core packaging
 * PortAudio network player backend:
   * PortAudio available in all platforms, including Linux.
   * Network player gets priority boost in Linux.
   * Discarded OSS API so in linux it uses Alsa (#Closes 200).
   * Xruns are notified on the console.
   * Available apis and devices are notified on the console.

Annotator/OnsetExtractor.py
#!/usr/bin/python


import sys
import os
from optparse import OptionParser

parser = OptionParser(
	usage="%prog [options] audiofile1 audiofile2 ...",
	version="%prog 1.0")
parser.add_option("-s","--print-schema",
		action="store_true",
		default=False,
		dest="printSchema",
		help="Outputs the description schema for the extractor"
	)
parser.add_option("-f",# "--suffix",
		default=".pool",
		dest="suffix",
		help="Appends SUFFIX to the generated descriptors file (default: '%default')"
	)
(options, args) = parser.parse_args()




if options.printSchema :
	print """<?xml version='1.0' encoding='UTF-8' standalone='no' ?>
<DescriptionScheme>

 <Uri>descriptionScheme:www.iua.upf.edu:clam:OnsetExtraction</Uri>

 <Attributes>
   <Attribute name='Onsets' scope='Song' type='Segmentation'>
     <ChildScope></ChildScope>
     <SegmentationPolicy>Unsized</SegmentationPolicy>
   </Attribute>
 </Attributes>

</DescriptionScheme>

"""
	sys.exit()

dataTemplate = """<?xml version='1.0' encoding='UTF-8' standalone='no' ?>
<Description>

  <ScopePool name="Song" size="1">
    <AttributePool name="Onsets" size="%i">
		%s
    </AttributePool>
  </ScopePool>

</Description>
"""

import sys
for audiofile in args:
	if audiofile.endswith(".ogg") :
		os.popen('ogg123 -d wav -f lala.wav %s'%audiofile).read()
		audiofile="lala.wav"
	if audiofile.endswith(".mp3") :
		os.popen('mpg123 -w lala.wav %s'%audiofile).read()
		audiofile="lala.wav"
	f=os.popen('./OnsetExtractor %s'%audiofile,'r')
	onsets = [ str(float(onset)/2) for onset in f ]
	target = open(audiofile+options.suffix,'w')
	print >> target, dataTemplate%(len(onsets), " ".join(onsets))
	target.close()




